pkg/controller/tailoredprofile/tailoredprofile_controller.go

```diff
@@ -187,6 +261,280 @@ func (r *ReconcileTailoredProfile) Reconcile(ctx context.Context, request reconc
+	if tp.Spec.Extends != "" {
+		if tp.Spec.DisableRules != nil || tp.Spec.EnableRules != nil || tp.Spec.ManualRules != nil || tp.Spec.SetValues != nil {
+			return true
+		}
+	} else {
+		if tp.Spec.EnableRules != nil || tp.Spec.ManualRules != nil {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *ReconcileTailoredProfile) handleDeprecation(
+	v1alphaTp *cmpv1alpha1.TailoredProfile, logger logr.Logger) (bool, error) {
+
+	deprecatedRules, err := r.getDeprecatedRules(v1alphaTp, logger)
+	if err != nil {
+		return false, err
+	}
+
+	deprecatedVariables, err := r.getDeprecatedVariables(v1alphaTp, logger)
+	if err != nil {
+		return false, err
+	}
+
+	doContinue := true
+	v1alphaTpCP := v1alphaTp.DeepCopy()
+
+	// remove any deprecated variables from the tailored profile
+	if len(v1alphaTp.Spec.SetValues) > 0 && len(deprecatedVariables) > 0 {
+		var newVariables []cmpv1alpha1.VariableValueSpec
+		for vi := range v1alphaTp.Spec.SetValues {
+			variables := &v1alphaTp.Spec.SetValues[vi]
+			if _, ok := deprecatedVariables[variables.Name]; ok {
+				doContinue = false
+				logger.Info("Removing deprecated variable", "variable", variables.Name)
+				r.Eventf(v1alphaTp, corev1.EventTypeWarning, "TailoredProfileDeprecatedSetting", "Removing deprecated variable: %s", variables.Name)
+				continue
+			} else {
+				newVariables = append(newVariables, *variables)
+			}
+		}
+		if len(newVariables) != len(v1alphaTp.Spec.SetValues) {
+			v1alphaTpCP.Spec.SetValues = newVariables
+		}
+	}
+
+	// remove any deprecated rules from the tailored profile
+	if len(v1alphaTp.Spec.EnableRules) > 0 && len(deprecatedRules) > 0 {
+		var newRules []cmpv1alpha1.RuleReferenceSpec
+		for ri := range v1alphaTp.Spec.EnableRules {
+			rule := &v1alphaTp.Spec.EnableRules[ri]
+			if _, ok := deprecatedRules[rule.Name]; ok {
+				doContinue = false
+				logger.Info("Removing deprecated rule from enableRules", "rule", rule.Name)
+				r.Eventf(v1alphaTp, corev1.EventTypeWarning, "TailoredProfileDeprecatedSetting", "Removing deprecated rule: %s", rule.Name)
+				continue
+			} else {
+				newRules = append(newRules, *rule)
+			}
+		}
+		if len(newRules) != len(v1alphaTp.Spec.EnableRules) {
+			v1alphaTpCP.Spec.EnableRules = newRules
+		}
+	}
+
+	if len(v1alphaTp.Spec.DisableRules) > 0 && len(deprecatedRules) > 0 {
+		var newRules []cmpv1alpha1.RuleReferenceSpec
+		for ri := range v1alphaTp.Spec.DisableRules {
+			rule := &v1alphaTp.Spec.DisableRules[ri]
+			if _, ok := deprecatedRules[rule.Name]; ok {
+				doContinue = false
+				logger.Info("Removing deprecated rule from disableRules", "rule", rule.Name)
+				r.Eventf(v1alphaTp, corev1.EventTypeWarning, "TailoredProfileDeprecatedSetting", "Removing deprecated rule: %s", rule.Name)
+				continue
+			} else {
+				newRules = append(newRules, *rule)
+			}
+		}
+		if len(newRules) != len(v1alphaTp.Spec.DisableRules) {
+			v1alphaTpCP.Spec.DisableRules = newRules
+		}
+	}
```

Comment:
I bet we could find a way to consolidate 402 - 437 into a single invocation that takes a rule list as input, and reuse it for both enabled and disabled rules.

Reply:
that's a good point, the reason why I did this way was to make it easier to see the logic on how we handling the migration.